engine-precompiles/src/xcc.rs

```diff
@@ -107,46 +115,101 @@ impl<I: IO> Precompile for CrossContractCall<I> {
 
         // It's not allowed to call cross contract call precompile in static or delegate mode
         if is_static {
-            return Err(ExitError::Other(Cow::from(consts::ERR_STATIC)));
+            return Err(revert_with_message(consts::ERR_STATIC));
         } else if context.address != cross_contract_call::ADDRESS.raw() {
-            return Err(ExitError::Other(Cow::from(consts::ERR_DELEGATE)));
+            return Err(revert_with_message(consts::ERR_DELEGATE));
         }
 
         let sender = context.caller;
         let target_account_id = create_target_account_id(sender, self.engine_account_id.as_ref());
         let args = CrossContractCallArgs::try_from_slice(input)
             .map_err(|_| ExitError::Other(Cow::from(consts::ERR_INVALID_INPUT)))?;
-        let promise = match args {
+        let (promise, attached_near) = match args {
             CrossContractCallArgs::Eager(call) => {
                 let call_gas = call.total_gas();
-                PromiseCreateArgs {
+                let attached_near = call.total_near();
+                let promise = PromiseCreateArgs {
                     target_account_id,
                     method: consts::ROUTER_EXEC_NAME.into(),
                     args: call
                         .try_to_vec()
                         .map_err(|_| ExitError::Other(Cow::from(consts::ERR_SERIALIZE)))?,
                     attached_balance: ZERO_YOCTO,
                     attached_gas: costs::ROUTER_EXEC + call_gas,
-                }
+                };
+                (promise, attached_near)
+            }
+            CrossContractCallArgs::Delayed(call) => {
+                let attached_near = call.total_near();
+                let promise = PromiseCreateArgs {
+                    target_account_id,
+                    method: consts::ROUTER_SCHEDULE_NAME.into(),
+                    args: call
+                        .try_to_vec()
+                        .map_err(|_| ExitError::Other(Cow::from(consts::ERR_SERIALIZE)))?,
+                    attached_balance: ZERO_YOCTO,
```

Comment:
It is possible for users to attach NEAR to their promises. They specify that in the input to the precompile. The engine itself does not ever attach any NEAR though because the users must pay for their own transactions from their router account. They can fund their router account using wNEAR inside the EVM (the purpose of this PR).

Reply:
Makes sense.